core/src/main/java/org/elasticsearch/cockroach/CockroachActionCoordinator.java

```diff
@@ -0,0 +1,264 @@
+                        if (newTask != null) {
+                            currentTasks.add(newTask);
+                        }
+                    } else {
+                        currentTasks.add(taskInProgress);
+                    }
+                }
+                if (found) {
+                    ClusterState.Builder builder = ClusterState.builder(currentState);
+                    return builder.putCustom(CockroachTasksInProgress.TYPE, new CockroachTasksInProgress(currentTasks)).build();
+                } else {
+                    logger.warn("Trying to {} cockroach task with id {} that no longer exists", source, uuid);
+                    return currentState;
+                }
+            }
+
+            @Override
+            public void onFailure(String source, Exception e) {
+                if (listener != null) {
+                    listener.onFailure(e);
+                }
+            }
+
+            @Override
+            public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
+                if (listener != null) {
+                    listener.onResponse(Empty.INSTANCE);
+                }
+            }
+
+        });
+
+    }
+
+    /**
+     * Removes a record about a running cockroach task from cluster state
+     *
+     * @param uuid     the uuid of a cockroach task
+     * @param listener the listener that will be called when task is removed
+     */
+    public void removeCockroachTask(String uuid, ActionListener<Empty> listener) {
+        performOperationOnCockroachTaskInProgress("remove cockroach task", uuid, taskInProgress -> null, listener);
+    }
+
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        if (event.localNodeMaster()) {
+            CockroachTasksInProgress tasks = event.state().custom(CockroachTasksInProgress.TYPE);
+            if (tasks != null && event.nodesRemoved()) {
+                // We need to check if removed nodes were running any of the tasks and reassign them
+                Set<String> removedNodes = event.nodesDelta().removedNodes().stream().map(DiscoveryNode::getId).collect(Collectors.toSet());
+                for (CockroachTaskInProgress taskInProgress : tasks.entries()) {
+                    if (removedNodes.contains(taskInProgress.getCallerNode())) {
+                        // The caller node disappeared, we need to assign a new caller node
+                        String action = taskInProgress.getAction();
+                        TransportCockroachAction cockroachAction = cockroachActionRegistry.getCockroachActionSafe(action);
+                        DiscoveryNode node = cockroachAction.responseNode(taskInProgress.getRequest(), event.state());
+                        if (node == null) {
+                            // no nodes are available to receive the response - just remove the task
+                            removeCockroachTask(taskInProgress.getUuid(), null);
```

Comment:
It is possible that while there isn't a response node to handle the response at the present time, the task could be still going on for a long time and a response node could be available at the time the task has completed, so this is prematurely removing the cockroach task.  Upon discussion with @imotov, this will be a future improvement and will be documented with a TODO